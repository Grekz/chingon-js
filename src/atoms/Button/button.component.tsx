import styled from '@emotion/styled'
import * as React from 'react'

interface IButtonProps {
  buttonType: ButtonTypes
  children?: React.ReactNode | string
  role?: string
}

type ButtonType = React.SFC<IButtonProps>

export enum ButtonTypes {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  SUCCESS = 'success',
  WARNING = 'warning',
  DANGER = 'danger',
  INFO = 'info',
}

const Button: ButtonType = props => (
  <ButtonWrapper role="button" buttonType={props.buttonType}>
    {props.children}
  </ButtonWrapper>
)

Button.defaultProps = {
  buttonType: ButtonTypes.PRIMARY,
  children: 'A simple button',
}

// Skins

interface ISkin {
  backgroundColor: string
  borderColor: string
  fontColor: string
}

const primary: ISkin = {
  backgroundColor: '#007bff',
  borderColor: '#007bff',
  fontColor: '#fff',
}

const primaryHover: ISkin = {
  backgroundColor: '#0069d9',
  borderColor: '#0062cc',
  fontColor: '#fff',
}

const secondary: ISkin = {
  backgroundColor: '#6c757d',
  borderColor: '#6c757d',
  fontColor: '#fff',
}

const secondaryHover: ISkin = {
  backgroundColor: '#5a6268',
  borderColor: '#545b62',
  fontColor: '#fff',
}

const success: ISkin = {
  backgroundColor: '#28a745',
  borderColor: '#28a745',
  fontColor: '#fff',
}

const successHover: ISkin = {
  backgroundColor: '#218838',
  borderColor: '#1e7e34',
  fontColor: '#fff',
}

const danger: ISkin = {
  backgroundColor: '#dc3545',
  borderColor: '#dc3545',
  fontColor: '#fff',
}

const dangerHover: ISkin = {
  backgroundColor: '#c82333',
  borderColor: '#bd2130',
  fontColor: '#fff',
}

const warning: ISkin = {
  backgroundColor: '#ffc107',
  borderColor: '#ffc107',
  fontColor: '#212529',
}
const warningHover: ISkin = {
  backgroundColor: '#e0a800',
  borderColor: '#d39e00',
  fontColor: '#212529',
}

const info: ISkin = {
  backgroundColor: '#17a2b8',
  borderColor: '#17a2b8',
  fontColor: '#fff',
}
const infoHover: ISkin = {
  backgroundColor: '#138496',
  borderColor: '#117a8b',
  fontColor: '#fff',
}

const getSkinProp = (
  buttonType: ButtonTypes,
  prop: string,
  isHover?: boolean
): string => {
  const { PRIMARY, SECONDARY, SUCCESS, WARNING, DANGER, INFO } = ButtonTypes
  switch (buttonType) {
    case PRIMARY:
      return isHover ? primaryHover[prop] : primary[prop]
    case SECONDARY:
      return isHover ? secondaryHover[prop] : secondary[prop]
    case SUCCESS:
      return isHover ? successHover[prop] : success[prop]
    case WARNING:
      return isHover ? warningHover[prop] : warning[prop]
    case DANGER:
      return isHover ? dangerHover[prop] : danger[prop]
    case INFO:
      return isHover ? infoHover[prop] : info[prop]
    default:
      return primary[prop]
  }
}

const ButtonWrapper: ButtonType = styled.button`
  cursor: pointer;
  display: inline-block;
  font-weight: 400;
  text-align: center;
  vertical-align: middle;
  user-select: none;
  background-color: transparent;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  color: ${({ buttonType }): string => getSkinProp(buttonType, 'fontColor')};
  background-color: ${({ buttonType }): string =>
    getSkinProp(buttonType, 'backgroundColor')};
  border-color: ${({ buttonType }): string =>
    getSkinProp(buttonType, 'borderColor')};
  &:hover {
    color: ${({ buttonType }): string =>
      getSkinProp(buttonType, 'fontColor', true)};
    background-color: ${({ buttonType }): string =>
      getSkinProp(buttonType, 'backgroundColor', true)};
    border-color: ${({ buttonType }): string =>
      getSkinProp(buttonType, 'borderColor', true)};
  }
`

export default Button
