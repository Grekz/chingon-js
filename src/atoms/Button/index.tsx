export { default, ButtonType } from './button.component'
