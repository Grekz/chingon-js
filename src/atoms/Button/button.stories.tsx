// @flow
import * as React from 'react'

// tslint:disable-next-line: no-implicit-dependencies
import { storiesOf } from '@storybook/react'
import Button from '.'

storiesOf('Atoms|Button', module).add('As Default', () => (
  <Button>Some button</Button>
))
