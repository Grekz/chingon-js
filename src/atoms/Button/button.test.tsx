import * as emotion from "emotion";
import { configure, shallow } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import serializer, { matchers } from "jest-emotion";
import * as React from "react";
import Button from ".";

configure({ adapter: new Adapter() });
expect.extend(matchers);
expect.addSnapshotSerializer(serializer);

describe("<Button /> rendering", () => {
  it("should render Button", () => {
    const wrapper = shallow(<Button />);
    expect(wrapper).toHaveLength(1);
  });
  it("hidden modal matches the snapshot", () => {
    const tree = shallow(<Button />);
    expect(toJson(tree)).toMatchSnapshot();
  });
  it("shown modal matches the snapshot", () => {
    const tree = shallow(<Button color={"white"} />);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
