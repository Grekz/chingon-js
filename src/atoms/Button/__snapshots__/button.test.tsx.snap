// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rendering hidden modal matches the snapshot 1`] = `
<Styled(button)
  buttonType="primary"
  role="button"
>
  A simple button
</Styled(button)>
`;

exports[`<Button /> rendering shown modal matches the snapshot 1`] = `
<Styled(button)
  buttonType="primary"
  role="button"
>
  A simple button
</Styled(button)>
`;
